src/algo/single_source/single_source.py

```diff
@@ -32,10 +34,49 @@ class SingleSourceAlgoSmart(SingleSourceSolver):
     def __init__(self, graph: LabelGraph, grammar: CnfGrammar):
         super().__init__(graph, grammar)
         self.index = SingleSourceIndex(graph, grammar)
+        self.index.init_simple_rules()
 
     def solve(self, sources_vertices: Iterable) -> Matrix:
-        # Use self.index
-        pass
+        # Initialize source matrices masks
+        for v in sources_vertices:
+            self.index.sources[self.index.grammar.start_nonterm][v, v] = True
+        # Algo's body
+        changed = True
+        while changed:
+            changed = False
+            # Iterate through all complex rules
+            for l, r1, r2 in self.index.grammar.complex_rules:
+
+                # Number of instances before operation
+                old_nnz = self.index.nonterms[l].nvals
+
+                # l -> r1 r2 ==> l += (l_src * r1) * r2 =>
+
+                # 1) r1_src += {(j, j) : (i, j) \in l_src}
+                i_l, j_l, vs_l = self.index.sources[l].to_lists()
+                for i, j in list(zip(i_l, j_l)):
+                    self.index.sources[r1][j, j] = True
```

Comment:
This can take quite a long time. There is another solution using reduce function to get vector of required j-th coordinates. So i uggest to implement it too as private method. After that we should compare both options.